When NRI plugin fails on RunPodSandbox, containerd leaks resources

The test for NRI failure leads to a leaked resource. Found while working on https://github.com/kubernetes-sigs/cri-tools/issues/2046

From AI:

  Root cause: Between StartSandbox succeeding (line 299) and the NRI RunPodSandbox hook (line 346), there is no defer to stop the sandbox/shim on failure. When NRI returns an error:                                                               
                                                                                                                                                                                                                                                    
  - Sandbox metadata is deleted from the store (defer at line 164) ✓                                                                                                                                                                                
  - NRI RemovePodSandbox notification is sent (defer at line 351) ✓                                                                                                                                                                                 
  - Shim task is NOT stopped — no call to sandboxService.StopSandbox() or controller.Stop() ✗                                                                                                                                                       
  - rootfs and shm mounts are NOT unmounted ✗                                                                                                                                                                                                       
                                                                                                                                                                                                                                                    
  The fix should add a defer between lines 313 and 344 that calls c.sandboxService.StopSandbox() when retErr != nil. The existing rollback tests (sandbox_run_rollback_test.go) don't cover NRI hook failure.

> But as mentioned there the `Tests passed but leaked resources detected — marking run as failed` is still thrown with or without the change I made so still looking into if that is because there are other leaks or because this change just isn't doing it right.

The check might not be ideal. Please take it with the grain of salt. My ultimate check was that CI should pass:

https://github.com/kubernetes-sigs/cri-tools/actions/runs/25479614094/job/74760625540?pr=2069

```
Cleanup temp directory /var/lib/containerd-critest created for cri tests
rm: cannot remove '/var/lib/containerd-critest/state/io.containerd.runtime.v2.task/k8s.io/7efca0ac6457d3cda74ce4ab8530e132be8486654062e74ee31a5927668fdc5e/rootfs': Device or resource busy
rm: cannot remove '/var/lib/containerd-critest/state/io.containerd.grpc.v1.cri/sandboxes/7efca0ac6457d3cda74ce4ab8530e132be8486654062e74ee31a5927668fdc5e/shm': Device or resource busy
Error: Process completed with exit code 1.
```

--- a/internal/cri/server/sandbox_run.go
+++ b/internal/cri/server/sandbox_run.go
@@ -312,6 +312,16 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		return nil, fmt.Errorf("failed to start sandbox %q: %w", id, err)
 	}
 
+	// Shutdown the sandbox if we fail before adding it to store.
+	rollbackSandbox := true
+	defer func() {
+		if retErr != nil && rollbackSandbox {
+			deferCtx, deferCancel := util.DeferContext()
+			defer deferCancel()
+			cleanupErr = c.sandboxService.ShutdownSandbox(deferCtx, sandbox.Sandboxer, id)
+		}
+	}()
+
 	if ctrl.Address != "" {
 		sandbox.Endpoint = sandboxstore.Endpoint{
 			Version: ctrl.Version,
@@ -368,6 +378,8 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 	if err := c.sandboxStore.Add(sandbox); err != nil {
 		return nil, fmt.Errorf("failed to add sandbox %+v into store: %w", sandbox, err)
 	}
+	// We no longer need to stop sandbox with a cleanup defer since it is in the store.
+	rollbackSandbox = false
 
 	// Send CONTAINER_CREATED event with both ContainerId and SandboxId equal to SandboxId.
 	// Note that this has to be done after sandboxStore.Add() because we need to get

